### ISSUE TITLE
tch::nn::Sgd constantly grows memory

### ISSUE THREAD

**Issue by jerry73204** [LaurentMazare/tch-rs#76]
I observed this issue when I tested scheduled lr with SGD in #68. After I switched from Adam to Sgd, the process memory grew over time and finally was killed by OOM. I profiled the binary using valgrind and compared Adam, SGD w and w/o `set_lr()`. The report indicates large memory consumption beneath `torch::optim::SGD::step()`. It occurs when SGD is used regardless of `set_lr()`.

The snapshot shows the foot page of heap memory using `valgrind --tool=massif`,
https://imgur.com/CK9ngdT

while this is Adam. It heap allocation keeps constant.
https://imgur.com/nLBleR7

**Comment by jerry73204** [comment]
Here provides the valgrind logs. Run `massif-visualizer FILE` to show fancy visualization.

http://wtf.csie.org/massif.out.adam
http://wtf.csie.org/massif.out.sgd

**Comment by LaurentMazare** [comment]
Interesting, thanks for reporting this.
Looking quickly at the code I don't see much difference in the way sgd and adam are handled. Maybe I'm missing something (or maybe there is an issue on the torch side?).
I also tried to reproduce things but didn't see much of a memory increase over time with sgd. Anyway I'll try digging deeper, if you have a small repro that could be quite helpful.

**Comment by LaurentMazare** [comment]
Here is a small repro using Adam. This leaks gpu memory very quickly for me.
Interestingly the leak disappears if you remove weight decay so my bet is that it's related to https://github.com/pytorch/pytorch/issues/20146 which hopefully should be fixed in 1.2.

```rust
extern crate tch;
use tch::nn::{FuncT, ModuleT, OptimizerConfig, SequentialT};
use tch::{Tensor, nn, Device, kind::Kind};

pub fn main() -> failure::Fallible<()> {
    let vs = nn::VarStore::new(Device::cuda_if_available());
    let net = nn::linear(vs.root(), 512, 10, Default::default());
    let mut opt = nn::Adam {
        beta1: 0.1,
        beta2: 0.1,
        wd: 0.0001,
    }
    .build(&vs, 0.0001)?;
    while true {
        let bimages = Tensor::zeros(&[64, 512], (Kind::Float, vs.device()));
        let loss = net
            .forward_t(&bimages, true)
            .mean();
        opt.backward_step(&loss);
    }
    Ok(())
}
```

**Comment by jerry73204** [comment]
I think you get the point. I checked the upstream [fix](https://github.com/pytorch/pytorch/pull/23007) and found the same patch should apply to others including Adam. It's time to send PR to pytorch.

**Comment by LaurentMazare** [comment]
Good catch! I incorrectly assumed that the original PR had fixed the issue for all optimizers...
I guess if you don't have lots of gpu memory we're a bit stuck on not using weight decay for now, it should not have that much of an impact on these trainings anyway.

**Comment by jerry73204** [comment]
I looked around if any distro has released patched binary for this issue. Saddly, no existing builds (pip, Ubuntu, Arch) have good luck.

For those struggle to get over with memory issue, I suggest to follow build instruction in my [patched PKGBUILD](https://gist.github.com/jerry73204/0d7efb335539434c43dfcd88db67c6f5). ([original PKGBUILD](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/python-pytorch)) Apply this [patch](https://gist.github.com/jerry73204/1fbd2f7a41a8eee1c47b163d40045454) before you start compiling. Finally pick up your `usr/{include,bin,share,lib}`.

During the process, I found the undocumented env `LIBTORCH_CXX11_ABI` pretty useful. The official binary is built with older ABI and torch-sys assumes `LIBTORCH_CXX11_ABI=0`. I did `export LIBTORCH_CXX11_ABI=1` manually such that torch-sys can link to my custom build. May we can write a doc for this someday.

**Comment by LaurentMazare** [comment]
The `LIBTORCH_CXX11_ABI` is indeed not much documented (sorry about that) and useful when you compile your own version of libtorch. I'll try to add something to the readme to reflect this.

**Comment by jerry73204** [comment]
I checked the 1.2.0 source and the patch is already merged. Now it can be closed.

**Comment by LaurentMazare** [comment]
Excellent, thanks for checking!
